build: upgrade bundled luarocks: 2.4.4 => 3.x

blueyed opened the pull request:
Changelog: https://github.com/luarocks/luarocks/compare/v2.4.4..v3.1.3#diff-4ac32a78649ca5bdd8e0ba38b7006a1e

LuaRocks 2.4.4 on Windows has:

> Created LuaRocks site-config file: C:/projects/nvim-deps/usr/luarocks\lua\luarocks\site_config_5_1.lua

> lpeg 1.0.1-1 is now installed in C:/projects/nvim-deps/usr

3.1.3:

> Created LuaRocks hardcoded settings file: C:/projects/nvim-deps/usr/luarocks\lua\luarocks\core\hardcoded.lua

> lpeg 1.0.1-1 is now installed in C:/projects/nvim-deps/usr/luarocks


The main problem with LuaRocks on AppVeyor/Windows seems to be that we're trying to force it installing .dlls into the dep's bin dir (`install.bat … /CMOD ${DEPS_BIN_DIR}`), but that does not seem to work anymore.

NOTES:

- `make install` with LuaRocks does not install 3.1.3 over 2.4.4 properly (https://github.com/luarocks/luarocks/issues/1038)

STATE:

- [ ] waiting for some feedback from LuaRocks upstream regarding the issues.  While we have workarounds, it seems good to wait for a stabilizing release, especially for Windows (where the workaround is necessargy on CI (env)).

blueyed committed:
build: upgrade bundled luarocks: 2.4.4 => 3.1.3

Changelog: https://github.com/luarocks/luarocks/compare/v2.4.4..v3.1.3#diff-4ac32a78649ca5bdd8e0ba38b7006a1e

blueyed commented:
Fails to install the luv rock on AppVeyor:
```
[119/122] cmd.exe /C "cd /D C:\projects\nvim-deps\build\src\luv && C:\projects\nvim-deps\usr\3.1\luarocks.bat make CC=C:/msys64/mingw64/bin/cc.exe LD=C:/msys64/mingw64/bin/cc.exe "CFLAGS=-O0 -g3 -fPIC" LIBUV_DIR=C:/projects/nvim-deps/usr LUA_COMPAT53_INCDIR=C:/projects/nvim-deps/build/src/lua-compat-5.3"
FAILED: usr/lib/luarocks/rocks/luv 
cmd.exe /C "cd /D C:\projects\nvim-deps\build\src\luv && C:\projects\nvim-deps\usr\3.1\luarocks.bat make CC=C:/msys64/mingw64/bin/cc.exe LD=C:/msys64/mingw64/bin/cc.exe "CFLAGS=-O0 -g3 -fPIC" LIBUV_DIR=C:/projects/nvim-deps/usr LUA_COMPAT53_INCDIR=C:/projects/nvim-deps/build/src/lua-compat-5.3"
src/luv.o: In function `luv_udp_connect':
C:\projects\nvim-deps\build\src\luv/src/udp.c:335: undefined reference to `uv_udp_connect'
src/luv.o: In function `luv_udp_getpeername':
C:\projects\nvim-deps\build\src\luv/src/udp.c:345: undefined reference to `uv_udp_getpeername'
src/luv.o: In function `luv_fs_opendir':
C:\projects\nvim-deps\build\src\luv/src/fs.c:731: undefined reference to `uv_fs_opendir'
src/luv.o: In function `luv_fs_readdir':
C:\projects\nvim-deps\build\src\luv/src/fs.c:741: undefined reference to `uv_fs_readdir'
src/luv.o: In function `luv_fs_closedir':
C:\projects\nvim-deps\build\src\luv/src/fs.c:754: undefined reference to `uv_fs_closedir'
src/luv.o: In function `luv_new_thread':
C:\projects\nvim-deps\build\src\luv/src/thread.c:327: undefined reference to `uv_thread_create_ex'
src/luv.o: In function `luv_os_uname':
C:\projects\nvim-deps\build\src\luv/src/misc.c:500: undefined reference to `uv_os_uname'
src/luv.o: In function `luv_gettimeofday':
C:\projects\nvim-deps\build\src\luv/src/misc.c:523: undefined reference to `uv_gettimeofday'
collect2.exe: error: ld returned 1 exit status
C:/msys64/mingw64/bin/cc.exe -O0 -g3 -fPIC -c -o src/luv.o -IC:/projects/nvim-deps/usr/include/luajit-2.0/ src/luv.c -IC:/projects/nvim-deps/usr/include -IC:/projects/nvim-deps/build/src/lua-compat-5.3
C:/msys64/mingw64/bin/cc.exe -shared -o luv.dll src/luv.o -LC:/projects/nvim-deps/usr -luv -lUser32 -lpsapi -liphlpapi -luserenv -lws2_32 -ladvapi32 C:/projects/nvim-deps/usr/lib/lua51.dll -lm
Error: Build error: Failed compiling module luv.dll
ninja: build stopped: subcommand failed.
DEBUG:  100+ cmake --build . --config $cmakeBuildType -- $cmakeGeneratorArgs ;  >>>> exitIfFailed
DEBUG:   22+ function exitIfFailed()  >>>> {
DEBUG:   23+   if ( >>>> $LastExitCode -ne 0) {
DEBUG:   24+      >>>> Set-PSDebug -Off
```

justinmk reviewed (dismissed):
"verbose" should be removed before committing to master. Our build logs are already too big.

blueyed commented:
> "verbose" should be removed before committing to master. Our build logs are already too big.

Sure, it's meant for debugging only - since it fails on AppVeyor.

blueyed commented:
luarocks 3 is broken with regard to luajit: https://github.com/luarocks/luarocks/pull/1042 - waiting for feedback there.  The hack/workaround is working on Travis, but not AppVeyor (but I've messed with install arguments there also).

justinmk commented on third-party/cmake/BuildLuarocks.cmake:
```diff
@@ -41,23 +41,23 @@ endfunction()
+# DEBUG
+list(APPEND LUAROCKS_BUILDARGS --verbose)
+
+if(MINGW OR UNIX OR (MINGW AND CMAKE_CROSSCOMPILING))
```
off-topic: I wonder if the `CMAKE_CROSSCOMPILING` cases can be removed since  b98b1168d81a0fd61c073780f0555ab858abdf78 and 55682710a

blueyed replied:
Hmm, good question.. no idea what it does really, and if it would be useful outside of Travis also?

justinmk replied:
commit message in https://github.com/neovim/neovim/commit/55682710a90ab5a70696dcca4567f640c30aab7d implies that cross-compiling doesn't work anyways. @equalsraf was the only one who was cross-compiling. So a quick review of the various `CMAKE_CROSSCOMPILING` sites is all that's needed to decide if it does anything useful.

blueyed commented:
Currently does not find lpeg on AppVeyor.
It gets copied by LuaRocks:

> s.execute: 	C: & cd "C:\Users\appveyor\AppData\Local\Temp\1\luarocks_lpeg-1.0.1-1-6049" & "C:/projects/nvim-deps/usr/luarocks/tools/cp.exe" "C:\projects\nvim-deps\usr\luarocks\lib\luarocks\rocks-5.1\lpeg\1.0.1-1\lib\lpeg.dll" "C:\projects\nvim-deps\usr\luarocks\lib\lua\5.1\lpeg.dll"

But that does not match the paths (CPATH) then:

```
============================
== LuaRocks is installed! ==
============================
You may want to add the following elements to your paths;
Lua interpreter;
  PATH     :   C:/projects/nvim-deps/usr/bin
  PATHEXT  :   .LUA
LuaRocks;
  PATH     :   C:/projects/nvim-deps/usr/luarocks
  LUA_PATH :   C:/projects/nvim-deps/usr/luarocks\lua\?.lua;C:/projects/nvim-deps/usr/luarocks\lua\?\init.lua
Local user rocktree (Note: %APPDATA% is user dependent);
  PATH     :   %APPDATA%\LuaRocks\bin
  LUA_PATH :   %APPDATA%\LuaRocks\share\lua\5.1\?.lua;%APPDATA%\LuaRocks\share\lua\5.1\?\init.lua
  LUA_CPATH:   %APPDATA%\LuaRocks\lib\lua\5.1\?.dll
System rocktree
  PATH     :   C:/projects/nvim-deps/usr/bin
  LUA_PATH :   C:/projects/nvim-deps/usr/bin/lua\?.lua;C:/projects/nvim-deps/usr/bin/lua\?\init.lua
  LUA_CPATH:   C:/projects/nvim-deps/usr/bin\?.dll
```

blueyed commented:
@erw7 said (in https://github.com/neovim/neovim/commit/828c9baad2a691ba87a99ee7dda7d7ac3b32cbe5#commitcomment-34064674):

> @blueyed In my environment, 3319e6d succeeds in building .deps. What if you set LUA_PATH and LUA_CPATH in 12938f1? Then, in my environment, I succeeded in building with MINGW_W64.

Good to know.
Are you using the bundled LuaJIT?
Do you set LUA_PATH and LUA_CPATH yourself?

How does lpeg behave for you?

blueyed commented:
@erw7 @janlazo 
It would be great if you could try this on Windows, based on the first commit (update), i.e. `make distclean ; make`.

erw7 commented:
@blueyed Finally I found a solution for the problem. I think that the problem that lpeg etc. are not found in the next patch will be solved.

```diff
diff --git a/third-party/cmake/BuildLuarocks.cmake b/third-party/cmake/BuildLuarocks.cmake
index 1f707ddd8..0bd16580f 100644
--- a/third-party/cmake/BuildLuarocks.cmake
+++ b/third-party/cmake/BuildLuarocks.cmake
@@ -91,7 +91,8 @@ elseif(MSVC OR MINGW)
     /SCRIPTS ${DEPS_BIN_DIR}
     /CMOD ${DEPS_BIN_DIR}
     ${COMPILER_FLAG}
-    /LUAMOD ${DEPS_BIN_DIR}/lua)
+    /LUAMOD ${DEPS_BIN_DIR}/lua
+    /CONFIG ${DEPS_INSTALL_DIR}/luarocks)

   set(LUAROCKS_BINARY ${DEPS_INSTALL_DIR}/luarocks/luarocks.bat)
 else()
```
`SYSCONFIDIR` is not output to `hardcoded.lua` unless `SYSCONFFORCE` on [`install.bat`#L1004](https://github.com/luarocks/luarocks/blob/3.1.3/install.bat#L1044) is `true`. `SYSCONFORCE` is `true` only if `/CONFIG` or `/SELFCONTAINED` is specified. Therefore, the directory to go to search for config file in [`set_confdirs()`](https://github.com/luarocks/luarocks/blob/3.1.3/src/luarocks/core/cfg.lua#L71-L87) was not set correctly. As a result, the rock was being installed in `${DEPS_INSTLL_DIR}/luarocks`  instead of `${DEPS_INSTALL_DIR}`.

erw7 commented:
@blueyed Build successfully with the first commit and the above patch (MINIGW_W64). I do not know whether it is necessary https://github.com/neovim/neovim/pull/10292/commits/12938f19ed4bee4e38a55aaf0f050a40059933d. The problem of undefined references to `uv_...` does not occur in my environment.

blueyed committed:
configure: --with-lua-interpreter=luajit

blueyed commented:
@erw7 
Awesome, thanks!

I've force-pushed minimal changes (without https://github.com/neovim/neovim/commit/12938f19ed4bee4e38a55aaf0f050a40059933d), and it  fails to luarocks-make luv now again.
How does it look for you?  Do you use `USE_BUNDLED_LUV` after all?

erw7 commented:
@blueyed For https://github.com/neovim/neovim/commit/12938f19ed4bee4e38a55aaf0f050a40059933d you are correct. The success of building `luv` in my environment was because `libuv.a` was present in `C:\msys64\mingw64\lib`.

blueyed commented:
ok.

We could specify LIBUV_LIBDIR explicitly I guess, but would rather like to avoid it, since it might indicate another issue in the first place.

It should be handled via https://github.com/luarocks/luarocks/blob/3e12f3df50e4f6f0035f13b255d380fbf207a7c1/src/luarocks/deps.lua#L421-L442 - can you debug it from there, e.g. by adding some print statements to see what is happening there?

erw7 commented:
@blueyed I understand the cause. The library key is missing in the external dependencies on rockspec. Therefore I think that `LIBUV_LIBDIR` variable was not created. Perhaps the behavior has changed with the luarocks version. The following changes are required to rockspec of luv:

```diff
diff --git a/rockspecs/luv-scm-0.rockspec b/rockspecs/luv-scm-0.rockspec
index 0aa2ee1..afa7f1a 100644
--- a/rockspecs/luv-scm-0.rockspec
+++ b/rockspecs/luv-scm-0.rockspec
@@ -23,7 +23,8 @@ dependencies = {

 external_dependencies = {
   LIBUV = {
-    header = 'uv.h'
+    header = 'uv.h',
+    library = 'uv'
   },
   LUA_COMPAT53 = {
     header = "c-api/compat-5.3.h"
```

Ref. https://github.com/luarocks/luarocks/wiki/Rockspec-format

blueyed commented:
@erw7 
Seems like only being required for windows/mingw though..

Related to https://github.com/luarocks/luarocks/blob/3e12f3df50e4f6f0035f13b255d380fbf207a7c1/src/luarocks/build/builtin.lua#L181

`add_flags` does the substitution: https://github.com/luarocks/luarocks/blob/3e12f3df50e4f6f0035f13b255d380fbf207a7c1/src/luarocks/build/builtin.lua#L160-L170

With

```diff
diff --git i/src/luarocks/build/builtin.lua w/src/luarocks/build/builtin.lua
index fb4431c..66c40a7 100644
--- i/src/luarocks/build/builtin.lua
+++ w/src/luarocks/build/builtin.lua
@@ -178,7 +178,9 @@ function builtin.run(rockspec)
       end
       compile_library = function(library, objects, libraries, libdirs)
          local extras = { unpack(objects) }
+         print("before", unpack(libdirs))
          add_flags(extras, "-L%s", libdirs)
+         print("after", unpack(libdirs))
          add_flags(extras, "-l%s", libraries)
          extras[#extras+1] = dir.path(variables.LUA_LIBDIR, variables.LUALIB)
          extras[#extras+1] = "-l" .. (variables.MSVCRT or "m")
@@ -241,7 +243,9 @@ function builtin.run(rockspec)
       end
       compile_library = function (library, objects, libraries, libdirs)
          local extras = { unpack(objects) }
+         print("before", unpack(libdirs))
          add_flags(extras, "-L%s", libdirs)
+         print("after", unpack(libdirs))
          if cfg.gcc_rpath then
             add_flags(extras, "-Wl,-rpath,%s", libdirs)
          end
```
I see (for the second case):
```
before  $(LIBUV_LIBDIR)
after   /home/daniel/Vcs/neovim/.deps/usr/lib
```

Can you check how it looks before/after for you?
"$(LIBUV_LIBDIR)" should be there, but appears to be not expanded properly.

blueyed commented:
There is something mingw specific in there: https://github.com/luarocks/luarocks/blob/3e12f3df50e4f6f0035f13b255d380fbf207a7c1/src/luarocks/build/builtin.lua#L160-L170

blueyed committed:
BuildLuarocks: use ROCKS_DIR, luarocks 3 appends the lua_version there

blueyed commented:
osx failure:
```
[8/15] Generating usr/lib/luarocks/rocks-5.1/mpack
FAILED: usr/lib/luarocks/rocks-5.1/mpack 
cd /Users/travis/nvim-deps && /Users/travis/nvim-deps/usr/bin/luarocks build mpack 1.0.7-0 CC=/Applications/Xcode-10.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc\ -isysroot/Applications/Xcode-10.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk LD=/Applications/Xcode-10.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc\ -isysroot/Applications/Xcode-10.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk
/Users/travis/nvim-deps/usr/bin/luajit: /Users/travis/nvim-deps/usr/share/lua/5.1/luarocks/cmd.lua:180: attempt to call field 'exists' (a nil value)
stack traceback:
	/Users/travis/nvim-deps/usr/share/lua/5.1/luarocks/cmd.lua:180: in function 'find_project_dir'
	/Users/travis/nvim-deps/usr/share/lua/5.1/luarocks/cmd.lua:259: in function 'detect_config_via_flags'
	/Users/travis/nvim-deps/usr/share/lua/5.1/luarocks/cmd.lua:267: in function 'init_config'
	/Users/travis/nvim-deps/usr/share/lua/5.1/luarocks/cmd.lua:373: in function 'run_command'
	/Users/travis/nvim-deps/usr/bin/luarocks:38: in main chunk
	[C]: at 0x0108309440
ninja: build stopped: subcommand failed.
```
(might be due to bad cache - I've seen `make distclean` being necessary with similar issues locally due to the upgrade - likely leaving old files behind then)

blueyed commented:
MSVC fails:
```
Project "C:\projects\nvim-deps\lpeg.vcxproj" (20) is building "C:\projects\nvim-deps\mpack.vcxproj" (21) on node 1 (default targets).
PrepareForBuild:
  Creating directory "Win32\RelWithDebInfo\mpack\".
  Creating directory "Win32\RelWithDebInfo\mpack\mpack.tlog\".
InitializeBuildStatus:
  Creating "Win32\RelWithDebInfo\mpack\mpack.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
ComputeCustomBuildOutput:
  Creating directory "C:\projects\nvim-deps\usr\lib\luarocks\rocks-5.1\".
CustomBuild:
  Generating usr/lib/luarocks/rocks-5.1/mpack
  'mingw32-gcc' is not recognized as an internal or external command,
  operable program or batch file.
  
Error : Build error : Failed compiling object lmpack.o [C:\projects\nvim-deps\mpack.vcxproj]
  
  mingw32-gcc -O2 -c -o lmpack.o -IC:/projects/nvim-deps/usr/include/luajit-2.0 lmpack.c
```

erw7 commented:
@blueyed I think it won't help much because it's late. https://github.com/neovim/neovim/pull/10292#issuecomment-506613368 It is as follows in MinGW.

```
before  $(LIBUV_LIBDIR)
after   C:/Users/erw7/src/neovim/.deps-test/usr
```

After all, if library is not specified in external_dependencies, [`check_external_dependency_at()`](https://github.com/luarocks/luarocks/blob/3.1.3/src/luarocks/deps.lua#L303-L388) will use the combination of prefix(`LIBUV_DIR`) and the first element of `cfg.external_deps_subdirs.lib`. The first element of `cfg.external_deps_subdirs.lib` is `lib` except for mingw32, but for mingw32 it is an empty string. If `LIBUV_LIBDIR` is defined, that takes precedence. Therefore, I think that the solution is good with https://github.com/neovim/neovim/pull/10292/commits/103f95855adbac3b089f289b69b2b51feddf2b4.

erw7 commented:
@blueyed https://github.com/neovim/neovim/pull/10292#issuecomment-506889725 Is not the environment variable `VCINSTALLDIR` defined?

Ref. https://github.com/luarocks/luarocks/blob/3.1.3/src/luarocks/core/cfg.lua#L608-L621

blueyed committed:
debug  [skip travis]

blueyed commented:
> Is not the environment variable `VCINSTALLDIR` defined?

Apparently not: https://ci.appveyor.com/project/neovim/neovim/builds/25635468

Includes:
```
VS110COMNTOOLS                 C:\Program Files (x86)\Microsoft Visual Studio 11.0\Common7\Tools\                      
VS120COMNTOOLS                 C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\Tools\                      
VS140COMNTOOLS                 C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\                      
VSSDK140Install                C:\Program Files (x86)\Microsoft Visual Studio 14.0\VSSDK\                              
```

I guess we could set it, but I would also expect SYSTEM to be Windows in hardcoded.lua in the first place.

blueyed committed:
TEST: set VCINSTALLDIR, test MINGW_64

erw7 commented:
> I guess we could set it, but I would also expect SYSTEM to be Windows in hardcoded.lua in the first place.

I also think that this fall back to `mingw32` is unnecessary. It is very difficult to understand that the error that `mingw32-gcc` can not be found occurs after installing with `MSVC`.

blueyed committed:
TEST: unset VCINSTALLDIR [skip travis]

blueyed committed:
update appveyor.yml [skip travis]

blueyed committed:
TODO/NOTE

blueyed committed:
TEST: https://github.com/luarocks/luarocks/pull/1051

blueyed committed:
BuildLuarocks.cmake: fix handling of non-bundled LuaJIT

blueyed committed:
Luarocks v3.2.1

blueyed commented:
Fails to install mpack now:

> /home/travis/nvim-deps/usr/bin/luajit: /home/travis/nvim-deps/usr/share/lua/5.1/luarocks/fs.lua:79: cfg is not initialized, please run cfg.init() first

```
./luarocks init
Initializing project 'luarocks' for Lua 5.1 ...
-----------------------------------------------
Checking your Lua installation ...
Adding entries to .gitignore ...
Preparing ./.luarocks/ ...
Wrote .luarocks/config-5.1.lua
Preparing ./lua_modules/ ...
./luarocks already exists. Not overwriting it!
Preparing ./lua for version 5.1...
mkdir -p "/home/travis/nvim-deps/usr/etc/luarocks"
install -m 644 "build/config-5.1.lua" "/home/travis/nvim-deps/usr/etc/luarocks/config-5.1.lua"
./luarocks make --tree="/home/travis/nvim-deps/usr"
No existing manifest. Attempting to rebuild...
luarocks 3.2.0-1 is now installed in /home/travis/nvim-deps/usr (license: MIT)
[81/90] Generating usr/lib/luarocks/rocks-5.1/mpack
FAILED: cd /home/travis/nvim-deps && /home/travis/nvim-deps/usr/bin/luarocks build mpack 1.0.8-0 CC=/usr/lib/ccache/cc LD=/usr/lib/ccache/cc
/home/travis/nvim-deps/usr/bin/luajit: /home/travis/nvim-deps/usr/share/lua/5.1/luarocks/fs.lua:79: cfg is not initialized, please run cfg.init() first
stack traceback:
	[C]: in function 'error'
	/home/travis/nvim-deps/usr/share/lua/5.1/luarocks/fs.lua:79: in function 'init'
	/home/travis/nvim-deps/usr/share/lua/5.1/luarocks/cmd.lua:447: in function 'run_command'
	...usr/lib/luarocks/rocks-5.1/luarocks/3.2.0-1/bin/luarocks:35: in main chunk
	[C]: at 0x00404560
ninja: build stopped: subcommand failed.
```

justinmk commented:
sourcehut failure looked network-related. Restarted it.

blueyed committed:
Revert "TODO/NOTE"

This reverts commit fe701152daafe1fb9beb0995948e48ed8f58ffdb.

blueyed committed:
appveyor.yml: remove DEBUG

blueyed commented:
First pass of cleaning up.
I guess it requires https://github.com/neovim/neovim/pull/11092 to be merged first (for https://github.com/luvit/luv/pull/353).

blueyed commented:
@justinmk 
> * Could remove `CMAKE_CROSSCOMPILING` [#10292 (comment)](https://github.com/neovim/neovim/pull/10292#discussion_r296500599)

Not really necessary, and I'd rather not touch it here - might be worth creating an issue though, since you've mentioned it before.

blueyed commented:
Tree-sitter fails to install on Appveyor here now?  /cc @bfredl 

```
DEBUG:  116+  >>>> cd c:\treesitter\tree-sitter-c-0.15.2
DEBUG:  117+  >>>> c:\treesitter\tree-sitter.exe test
Failed to execute C compiler
Details:
  The system cannot find the file specified. (os error 2)
```
https://ci.appveyor.com/project/neovim/neovim/builds/27763787/job/tbw0mxfs3cq4vh1s?fullLog=true

blueyed committed:
Only set VCINSTALLDIR with MSVC, after tree-sitter setup

Set VCINSTALLDIR after tree-sitter.exe-test.
Ref: https://github.com/tree-sitter/tree-sitter-c/issues/33

Update VCINSTALLDIR based on CMake's output.  Not really necessary, but
needs to be moved anyway.

blueyed commented:
The problem with MSVC and "tree-sitter.exe test":
```
DEBUG:  119+  >>>> c:\treesitter\tree-sitter.exe --help
tree-sitter 0.15.9 (0795b59695854505e74e1dd70f48bbbd1a354414)
Max Brunsfeld <maxbrunsfeld@gmail.com>
Generates and tests parsers
USAGE:
    tree-sitter.exe <SUBCOMMAND>
FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information
SUBCOMMANDS:
    build-wasm        Compile a parser to WASM
    dump-languages    Print info about all known language parsers
    generate          Generate a parser
    help              Prints this message or the help of the given subcommand(s)
    highlight         Highlight a file
    init-config       Generate a default config file
    parse             Parse files
    query             Search files using a syntax tree query
    test              Run a parser's tests
    web-ui            Test a parser interactively in the browser
DEBUG:  120+  >>>> c:\treesitter\tree-sitter.exe test
Failed to execute C compiler
Details:
  The system cannot find the file specified. (os error 2)
```
Trying a correct `$VCINSTALLDIR` now.

blueyed committed:
TEST: Travis: CACHE_ENABLE=false  [skip appveyor]

Check if it fixes:

    Using third-party dependencies from Travis cache (last update: 2019-10-09 02:03:42.000000000 +0000).
    Configuring with '-DUSE_BUNDLED_GPERF=OFF -DUSE_BUNDLED_LUA=ON'.
    -- Found GNU Make at /usr/bin/make
    -- Lua target is linux
    -- Configuring done
    -- Generating done
    -- Build files have been written to: /home/travis/nvim-deps
    ==============================================================================
    [4/16] Generating usr/lib/luarocks/rocks-5.1/luv
    FAILED: cd /home/travis/nvim-deps/build/src/luv && /home/travis/nvim-deps/usr/bin/luarocks make CC=/usr/lib/ccache/cc LD=/usr/lib/ccache/cc CFLAGS=-O0\ -g3\ -fPIC LIBUV_DIR=/home/travis/nvim-deps/usr LUA_COMPAT53_INCDIR=/home/travis/nvim-deps/build/src/lua-compat-5.3
    /home/travis/nvim-deps/usr/bin/luajit: ...usr/lib/luarocks/rocks-5.1/luarocks/3.2.1-1/bin/luarocks:4: module 'luarocks.core.cfg' not found:
    	no field package.preload['luarocks.core.cfg']
    	no file '/home/travis/nvim-deps/usr/share/lua/5.1/luarocks/core/cfg.lua'
    	no file '/home/travis/nvim-deps/usr/share/lua/5.1/luarocks/core/cfg/init.lua'
    	no file './luarocks/core/cfg.lua'
    	no file '/home/travis/nvim-deps/usr/share/luajit-2.1.0-beta3/luarocks/core/cfg.lua'
    	no file '/usr/local/share/lua/5.1/luarocks/core/cfg.lua'
    	no file '/usr/local/share/lua/5.1/luarocks/core/cfg/init.lua'
    	no file '/home/travis/nvim-deps/usr/share/lua/5.1/luarocks/core/cfg.lua'
    	no file '/home/travis/nvim-deps/usr/share/lua/5.1/luarocks/core/cfg/init.lua'
    	no file '/home/travis/nvim-deps/usr/lib/lua/5.1/luarocks/core/cfg.so'
    	no file './luarocks/core/cfg.so'
    	no file '/usr/local/lib/lua/5.1/luarocks/core/cfg.so'
    	no file '/home/travis/nvim-deps/usr/lib/lua/5.1/luarocks/core/cfg.so'
    	no file '/usr/local/lib/lua/5.1/loadall.so'
    	no file '/home/travis/nvim-deps/usr/lib/lua/5.1/luarocks.so'
    	no file './luarocks.so'
    	no file '/usr/local/lib/lua/5.1/luarocks.so'
    	no file '/home/travis/nvim-deps/usr/lib/lua/5.1/luarocks.so'
    	no file '/usr/local/lib/lua/5.1/loadall.so'
    stack traceback:
    	[C]: in function 'require'
    	...usr/lib/luarocks/rocks-5.1/luarocks/3.2.1-1/bin/luarocks:4: in main chunk
    	[C]: at 0x00404c50

justinmk commented:
@blueyed this is marked RDY, any reason it wasn't merged?  perhaps rebase?

justinmk commented on third-party/cmake/BuildLuarocks.cmake:
```diff
@@ -52,23 +52,41 @@ if(NOT MSVC)
 
+# Lua version, used with rocks directories.
+# Defaults to 5.1 for bundled LuaJIT/Lua.
+set(lua_version "5.1")
```
why not uppercase?

```suggestion
set(LUA_VERSION "5.1")
```

blueyed replied:
Changed to uppercase (also others).
IIRC this was an attempt to distinguish "internal"/local vars from others.

blueyed commented:
@justinmk 
Rebased for now.  IIRC it needs cleaning up, and is still not ready completely.

blueyed committed:
Revert "TEST: Travis: CACHE_ENABLE=false  [skip appveyor]"

This reverts commit b3041e54e3de113098d6983308cf7e296467bbc3.

blueyed committed:
uppercase vars

blueyed commented:
Looks good now, will squash-merge when CI passes.
